PermanentOrg/web-app
# Upgrade to Angular 16

This PR upgrades the web-app's Angular version to 16, and upgrades all additional dependencies that need to be updated. Compared to previous Angular updates this is comparatively simpler update with a few small notes to watch out for.

- The only automatic code migration performed by the Angular CLI was the removal of the `CanActivate` interface from code.
- This upgrade caused compatibility to break for certain legacy packages. Thankfully, all packages are still maintained so we needed to just upgrade major versions, and because of the gradual updates we've done over time, most packages we're using were already upgraded to be compatible with Angular 16.
  - There was **one** exception to this rule, which was the [ngx-route-history](https://github.com/andrewatwood/ngx-route-history) package which is a small package created by Andrew (previous developer) back in 2019. Since this package was so small, it made sense to just reimplement this functionality in the web-app. See commit history for more details on this process.

**Steps to Test**
1. Run `npm install`
2. Run `npm run build` / `npm run build:local` to verify that a production build can be made without errors.
3. Run unit tests to verify that they can be run without errors and no functionality covered by tests is broken.
4. Run `npm run storybook` to verify Storybook can be run without errors <small>(I always forget this one!)</small>
5. Run `npm run dev` to verify that the local development server can be run without errors
  - Click around in the web-app to make sure nothing's broken!

Resolves #317. Resolves PER-9399.
